File: json/json-annotation-processor/src/main/java/ru/tinkoff/kora/json/annotation/processor/extension/JsonKoraExtension.java
package ru.tinkoff.kora.json.annotation.processor.extension;

import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;
import ru.tinkoff.kora.json.annotation.processor.JsonProcessor;
import ru.tinkoff.kora.json.annotation.processor.JsonTypes;
import ru.tinkoff.kora.json.annotation.processor.JsonUtils;
import ru.tinkoff.kora.json.annotation.processor.KnownType;
import ru.tinkoff.kora.json.annotation.processor.reader.ReaderTypeMetaParser;
import ru.tinkoff.kora.json.annotation.processor.writer.WriterTypeMetaParser;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionResult;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.Nullable;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.Objects;
import java.util.Set;

public class JsonKoraExtension implements KoraExtension {

    private final Types types;
    private final Elements elements;
    private final JsonProcessor processor;
    private final TypeMirror jsonWriterErasure;
    private final TypeMirror jsonReaderErasure;
    private final ReaderTypeMetaParser readerTypeMetaParser;
    private final WriterTypeMetaParser writerTypeMetaParser;

    public JsonKoraExtension(ProcessingEnvironment processingEnv) {
        this.types = processingEnv.getTypeUtils();
        this.elements = processingEnv.getElementUtils();
        var knownTypes = new KnownType();
        this.readerTypeMetaParser = new ReaderTypeMetaParser(processingEnv, knownTypes);
        this.writerTypeMetaParser = new WriterTypeMetaParser(processingEnv, knownTypes);

        this.processor = new JsonProcessor(processingEnv);
        this.jsonWriterErasure = this.types.erasure(this.elements.getTypeElement(JsonTypes.jsonWriter.canonicalName()).asType());
        this.jsonReaderErasure = this.types.erasure(this.elements.getTypeElement(JsonTypes.jsonReader.canonicalName()).asType());
    }

    @Override
    public KoraExtensionDependencyGenerator getDependencyGenerator(RoundEnvironment roundEnvironment, TypeMirror typeMirror, Set<String> tags) {
        if (!tags.isEmpty()) return null;
        var erasure = this.types.erasure(typeMirror);
        if (this.types.isSameType(erasure, this.jsonWriterErasure)) {
            var writerType = (DeclaredType) typeMirror;
            var possibleJsonClass = writerType.getTypeArguments().get(0);
            if (possibleJsonClass.getKind() != TypeKind.DECLARED) {
                return null;
            }
            var jsonElement = (TypeElement) this.types.asElement(possibleJsonClass);
            if (AnnotationUtils.findAnnotation(jsonElement, JsonTypes.json) != null || AnnotationUtils.findAnnotation(jsonElement, JsonTypes.jsonWriterAnnotation) != null) {
                return KoraExtensionDependencyGenerator.generatedFrom(elements, jsonElement, "JsonWriter");
            }
            if (jsonElement.getModifiers().contains(Modifier.SEALED) || jsonElement.getKind() == ElementKind.ENUM) {
                return () -> this.generateWriter(possibleJsonClass);
            }
            try {
                Objects.requireNonNull(this.writerTypeMetaParser.parse(jsonElement, possibleJsonClass));
                return () -> this.generateWriter(possibleJsonClass);
            } catch (ProcessingErrorException e) {
                return null;
            }
        }
        if (this.types.isSameType(erasure, this.jsonReaderErasure)) {
            var readerType = (DeclaredType) typeMirror;
            var possibleJsonClass = readerType.getTypeArguments().get(0);
            if (possibleJsonClass.getKind() != TypeKind.DECLARED) {
                return null;
            }
            var typeElement = (TypeElement) types.asElement(possibleJsonClass);
            if (AnnotationUtils.findAnnotation(typeElement, JsonTypes.json) != null
                || AnnotationUtils.findAnnotation(typeElement, JsonTypes.jsonReaderAnnotation) != null
                || CommonUtils.findConstructors(typeElement, s -> s.contains(Modifier.PUBLIC))
                    .stream()
                    .anyMatch(e -> AnnotationUtils.findAnnotation(e, JsonTypes.jsonReaderAnnotation) != null)) {
                return KoraExtensionDependencyGenerator.generatedFrom(elements, typeElement, "JsonReader");
            }

            if (typeElement.getModifiers().contains(Modifier.SEALED) || typeElement.getKind() == ElementKind.ENUM) {
                return () -> this.generateReader(possibleJsonClass);
            }
            try {
                Objects.requireNonNull(this.readerTypeMetaParser.parse(typeElement, typeMirror));
                return () -> this.generateReader(possibleJsonClass);
            } catch (ProcessingErrorException e) {
                return null;
            }
        }
        return null;
    }


    @Nullable
    private ExtensionResult generateReader(TypeMirror jsonClass) {
        var jsonTypeElement = (TypeElement) this.types.asElement(jsonClass);
        var packageElement = this.elements.getPackageOf(jsonTypeElement).getQualifiedName().toString();
        var resultClassName = JsonUtils.jsonReaderName(this.types, jsonClass);
        var resultElement = this.elements.getTypeElement(packageElement + "." + resultClassName);
        if (resultElement != null) {
            return buildExtensionResult(resultElement);
        }

        var hasJsonConstructor = CommonUtils.findConstructors(jsonTypeElement, s -> !s.contains(Modifier.PRIVATE))
            .stream()
            .anyMatch(e -> AnnotationUtils.findAnnotation(e, JsonTypes.jsonReaderAnnotation) != null);
        if (hasJsonConstructor || AnnotationUtils.findAnnotation(jsonTypeElement, JsonTypes.jsonReaderAnnotation) != null) {
            // annotation processor will handle that
            return ExtensionResult.nextRound();
        }

        this.processor.generateReader(jsonTypeElement);
        return ExtensionResult.nextRound();
    }

    @Nullable
    private ExtensionResult generateWriter(TypeMirror jsonClass) {
        var jsonTypeElement = (TypeElement) this.types.asElement(jsonClass);
        var packageElement = this.elements.getPackageOf(jsonTypeElement).getQualifiedName().toString();
        var resultClassName = JsonUtils.jsonWriterName(this.types, jsonClass);
        var resultElement = this.elements.getTypeElement(packageElement + "." + resultClassName);
        if (resultElement != null) {
            return buildExtensionResult(resultElement);
        }

        if (AnnotationUtils.findAnnotation(jsonTypeElement, JsonTypes.json) != null || AnnotationUtils.findAnnotation(jsonTypeElement, JsonTypes.jsonWriterAnnotation) != null) {
            // annotation processor will handle that
            return ExtensionResult.nextRound();
        }

        this.processor.generateWriter(jsonTypeElement);
        return ExtensionResult.nextRound();
    }

    private ExtensionResult buildExtensionResult(TypeElement resultElement) {
        var constructor = findDefaultConstructor(resultElement);

        return ExtensionResult.fromExecutable(constructor);
    }


    private ExecutableElement findDefaultConstructor(TypeElement resultElement) {
        return resultElement.getEnclosedElements()
            .stream()
            .filter(e -> e.getKind() == ElementKind.CONSTRUCTOR)
            .map(ExecutableElement.class::cast)
            .findFirst()
            .orElseThrow();
    }
}

File: json/json-symbol-processor/src/main/kotlin/ru/tinkoff/kora/json/ksp/extension/JsonKoraExtension.kt
package ru.tinkoff.kora.json.ksp.extension

import com.google.devtools.ksp.*
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.*
import ru.tinkoff.kora.json.ksp.*
import ru.tinkoff.kora.json.ksp.reader.ReaderTypeMetaParser
import ru.tinkoff.kora.json.ksp.writer.WriterTypeMetaParser
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension
import ru.tinkoff.kora.ksp.common.AnnotationUtils.isAnnotationPresent
import ru.tinkoff.kora.ksp.common.KspCommonUtils.parametrized
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException

class JsonKoraExtension(
    private val resolver: Resolver,
    private val kspLogger: KSPLogger,
    codeGenerator: CodeGenerator
) : KoraExtension {
    private val jsonWriterErasure = resolver.getClassDeclarationByName(JsonTypes.jsonWriter.canonicalName)!!.asStarProjectedType()
    private val jsonReaderErasure = resolver.getClassDeclarationByName(JsonTypes.jsonReader.canonicalName)!!.asStarProjectedType()
    private val knownTypes = KnownType(resolver)
    private val readerTypeMetaParser: ReaderTypeMetaParser = ReaderTypeMetaParser(knownTypes, kspLogger)
    private val writerTypeMetaParser: WriterTypeMetaParser = WriterTypeMetaParser(resolver)
    private val processor: JsonProcessor = JsonProcessor(resolver, kspLogger, codeGenerator, knownTypes)

    override fun getDependencyGenerator(resolver: Resolver, type: KSType): (() -> ExtensionResult)? {
        val actualType = type.makeNotNullable()
        val erasure = actualType.starProjection()
        if (erasure == jsonWriterErasure) {
            val possibleJsonClass = type.arguments[0].type!!.resolve()
            if (possibleJsonClass.isMarkedNullable) {
                val jsonWriterDecl = resolver.getClassDeclarationByName(JsonTypes.jsonWriter.canonicalName)!!
                val functionDecl = resolver.getFunctionDeclarationsByName("ru.tinkoff.kora.json.common.JsonKotlin.writerForNullable").first()
                val writerType = jsonWriterDecl.asType(
                    listOf(
                        resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(possibleJsonClass), Variance.INVARIANT)
                    )
                )
                val delegateType = jsonWriterDecl.asType(
                    listOf(
                        resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(possibleJsonClass.makeNotNullable()), Variance.INVARIANT)
                    )
                )
                val functionType = functionDecl.parametrized(writerType, listOf(delegateType))
                return { ExtensionResult.fromExecutable(functionDecl, functionType) }
            }
            val possibleJsonClassDeclaration = possibleJsonClass.declaration
            if (possibleJsonClassDeclaration !is KSClassDeclaration) {
                return null
            }
            if (possibleJsonClassDeclaration.isAnnotationPresent(JsonTypes.json) || possibleJsonClassDeclaration.isAnnotationPresent(JsonTypes.jsonWriter)) {
                return generatedByProcessor(resolver, possibleJsonClassDeclaration, "JsonWriter")
            }
            if (possibleJsonClassDeclaration.modifiers.contains(Modifier.ENUM) || possibleJsonClassDeclaration.modifiers.contains(Modifier.SEALED)) {
                return { generateWriter(resolver, possibleJsonClassDeclaration) }
            }
            try {
                writerTypeMetaParser.parse(possibleJsonClassDeclaration)
                return { generateWriter(resolver, possibleJsonClassDeclaration) }
            } catch (e: ProcessingErrorException) {
                e.message?.let { kspLogger.warn(it, null) }
                return null
            }
        }
        if (erasure == jsonReaderErasure) {
            val possibleJsonClass = type.arguments[0].type!!.resolve()
            if (possibleJsonClass.isMarkedNullable) {
                val jsonReaderDecl = resolver.getClassDeclarationByName(JsonTypes.jsonReader.canonicalName)!!
                val functionDecl = resolver.getFunctionDeclarationsByName("ru.tinkoff.kora.json.common.JsonKotlin.readerForNullable").first()
                val readerType = jsonReaderDecl.asType(
                    listOf(
                        resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(possibleJsonClass), Variance.INVARIANT)
                    )
                )
                val delegateType = jsonReaderDecl.asType(
                    listOf(
                        resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(possibleJsonClass.makeNotNullable()), Variance.INVARIANT)
                    )
                )
                val functionType = functionDecl.parametrized(readerType, listOf(delegateType))
                return { ExtensionResult.fromExecutable(functionDecl, functionType) }
            }
            val possibleJsonClassDeclaration = possibleJsonClass.declaration
            if (possibleJsonClassDeclaration !is KSClassDeclaration) {
                return null
            }
            if (possibleJsonClassDeclaration.isAnnotationPresent(JsonTypes.json)
                || possibleJsonClassDeclaration.isAnnotationPresent(JsonTypes.jsonReader)
                || possibleJsonClassDeclaration.primaryConstructor?.isAnnotationPresent(JsonTypes.jsonReader) == true
            ) {
                return generatedByProcessor(resolver, possibleJsonClassDeclaration, "JsonReader")
            }
            if (possibleJsonClassDeclaration.modifiers.contains(Modifier.ENUM) || possibleJsonClassDeclaration.modifiers.contains(Modifier.SEALED)) {
                return { generateReader(resolver, possibleJsonClassDeclaration) }
            }

            try {
                readerTypeMetaParser.parse(possibleJsonClassDeclaration)
                return { generateReader(resolver, possibleJsonClassDeclaration) }
            } catch (e: ProcessingErrorException) {
                return null
            }
        }
        return null
    }

    private fun generateReader(resolver: Resolver, jsonClass: KSClassDeclaration): ExtensionResult {
        val packageElement = jsonClass.packageName.asString()
        val resultClassName = jsonClass.jsonReaderName()
        val resultDeclaration = resolver.getClassDeclarationByName("$packageElement.$resultClassName")
        if (resultDeclaration != null) {
            return ExtensionResult.fromConstructor(findDefaultConstructor(resultDeclaration), resultDeclaration)
        }
        val hasJsonConstructor = jsonClass.getConstructors().filter { !it.isPrivate() }.any { it.isAnnotationPresent(JsonTypes.jsonReaderAnnotation) }
        if (hasJsonConstructor || jsonClass.isAnnotationPresent(JsonTypes.jsonReaderAnnotation)) {
            // annotation processor will handle that
            return ExtensionResult.RequiresCompilingResult
        }
        processor.generateReader(jsonClass)
        return ExtensionResult.RequiresCompilingResult
    }

    private fun generateWriter(resolver: Resolver, declaration: KSClassDeclaration): ExtensionResult {
        val packageElement = declaration.packageName.asString()
        val resultClassName = declaration.jsonWriterName()
        val resultDeclaration = resolver.getClassDeclarationByName("$packageElement.$resultClassName")
        if (resultDeclaration != null) {
            return ExtensionResult.fromConstructor(findDefaultConstructor(resultDeclaration), resultDeclaration)
        }
        if (declaration.isAnnotationPresent(JsonTypes.json) || declaration.isAnnotationPresent(JsonTypes.jsonWriterAnnotation)) {
            // annotation processor will handle that
            return ExtensionResult.RequiresCompilingResult
        }
        processor.generateWriter(declaration)
        return ExtensionResult.RequiresCompilingResult
    }

    private fun findDefaultConstructor(resultElement: KSClassDeclaration): KSFunctionDeclaration {
        return resultElement.primaryConstructor ?: throw NoSuchElementException("No value present")
    }

    private fun isProcessableType(typeRef: KSType): Boolean {
        val declaration = typeRef.declaration as KSClassDeclaration
        val classKind = declaration.classKind
        return classKind == ClassKind.ENUM_CLASS || (classKind == ClassKind.CLASS && !typeRef.declaration.isOpen())
    }
}
